modules/dnn/src/cuda/shortcut.cu

```diff
@@ -0,0 +1,108 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include <cuda_runtime.h>
+#include <cuda_fp16.h>
+
+#include "grid_stride_range.hpp"
+#include "execution.hpp"
+#include "vector_traits.hpp"
+
+#include "../cuda4dnn/csl/stream.hpp"
+#include "../cuda4dnn/csl/span.hpp"
+#include "../cuda4dnn/csl/tensor.hpp"
+
+#include <opencv2/core.hpp>
+
+using namespace cv::dnn::cuda4dnn::csl;
+using namespace cv::dnn::cuda4dnn::csl::device;
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace kernels {
+
+    namespace raw {
+        template <class T, std::size_t N>
+        __global__ void input_shortcut_vec(
+            Span<T> output,
+            View<T> input, index_type c_input, /* `c_input` = number of channels in `input` */
+            View<T> from, index_type c_from, /* `c_from` = number of channels in `from` */
+            size_type channel_stride /* common for both `input` and `from` */
+        ) {
+            using vector_type = get_vector_type_t<T, N>;
+
+            auto output_vPtr = vector_type::get_pointer(output.data());
+            auto input_vPtr = vector_type::get_pointer(input.data());
+            auto from_vPtr = vector_type::get_pointer(from.data());
+
+            auto batch_stride_input = c_input * channel_stride;
+            auto batch_stride_from = c_from * channel_stride;
+
+            for (auto i : grid_stride_range(output.size() / vector_type::size())) {
+                const auto actual_idx = i * vector_type::size();
+                const auto b = actual_idx / batch_stride_input; /* `input` and `output` have the same shape */
+                const auto c = (actual_idx % batch_stride_input) / channel_stride;
+
+                vector_type vec_input;
+                v_load(vec_input, input_vPtr[i]);
+
+                /* We can break down the shortcut operation into two steps:
+                 * - copy `input` to `output`
+                 * - add `from` to corresponding channels in `output`
+                 *
+                 * In this scheme, only some channels in the `output` differ from `input`. They differ in the channels
+                 * which have a corresponding channel in `from`.
+                 */
+                if (c < c_from) {
+                    const auto from_actual_idx = b * batch_stride_from + c * channel_stride;
+                    const auto from_vec_idx = from_actual_idx / vector_type::size();
+
+                    vector_type vec_from;
+                    v_load(vec_from, from_vPtr[from_vec_idx]);
```

Comment:
If I change `from_vPtr[from_vec_idx]` to `from_vPtr[i]`, OCV CPU backend and the CUDA backend agree with their outputs. But I have doubts over which one is correct.

I used `input_shortcut_kernel` as reference for this PR:

https://github.com/AlexeyAB/darknet/blob/eac26226a7fc0a9da2b684a564f8f086eaf38390/src/blas_kernels.cu#L714-L729

```
int out_index = i*sample + w2*(j*sample + h2*(k + c2*b));
int add_index = i*stride + w1*(j*stride + h1*(k + c1*b));
out[out_index] = in[out_index] + add[add_index];
```
which is
```
int out_index = i + w*(j + h*(k + c2*b));
    int add_index = i + w*(j + h*(k + c1*b));
    out[out_index] = in[out_index] + add[add_index];
```
if the spatial dimensions are same.

As it can be seen, the values in the corresponding channels of both inputs are added. I can infer the same from https://github.com/opencv/opencv/pull/15739#issuecomment-544036806. This is what I think I have done in this PR.

The OpenCV CPU code does something different which I think is wrong and can potentially lead to segfaults.

The eltwise layer code in OCV backend assumes that the number of channels are same in all inputs. `globalDelta` is calculated [here](https://github.com/opencv/opencv/pull/15739/files#diff-7ac73ff12c29882cb913b6f09da2f82cR237). The same `globalDelta` is also used here for indexing another input [here for example](https://github.com/opencv/opencv/pull/15739/files#diff-7ac73ff12c29882cb913b6f09da2f82cR290). But two inputs have different number of channels and hence, the same `globalDelta` is not applicable. If the second input had lesser number of channels than for which `globalDelta` was calculated for, it could lead to an illegal memory access in the second input.

Is this a bug in the OCV backend or I have misunderstood what the shortcut layer does?

/cc @dkurt

Reply:
I tried running the object detection sample on both variants. Unfortunately, both give reasonable outputs and hence it isn't obvious which one is correct.

Reply:
Perhaps we need some test for this single layer only.

Reply:
Silly me. The two variants are identical when batch size is one which was the case in the object detection sample. When I run a batch of 10 images, the CUDA and OCV backend agree on the first image only as expected.

Reply:
I did some tests with two images in a batch.

https://gist.github.com/YashasSamaga/b895dc109d50cbbb4e171c5d8e64defd

The CUDA backend agrees with the darknet output while the CPU backend does not. So it's a bug in the OpenCV CPU backend.